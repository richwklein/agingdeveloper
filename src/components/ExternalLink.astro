---
/**
 * Component used to render a link to an external site.
 *
 * @param {Props} props - The properties of the link.
 * @return {Function} an Astro component factory function.
 *
 * @example
 * <ExternalLink to="https://github.com/richwklein">Github</ExternalLink>
 */
interface Props {
  to: string;
  class?: string;
}

const { to, class: className } = Astro.props;
---

<!--prettier-ignore -->
<a data-link="external" target="_blank" rel="noopener noreferrer" href={to} class:list={[className]}><slot /></a>
