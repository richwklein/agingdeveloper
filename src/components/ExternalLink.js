import React from 'react';
import {Link} from '@material-ui/core';
import {OutboundLink} from 'gatsby-plugin-google-analytics';

const ForwardOutboundLink = React.forwardRef((props, ref) => (
  <OutboundLink {...props} />
));

const ExternalLink = ({to, children}) => {
  return (
    <Link
      component={ForwardOutboundLink}
      href={to}
      target="_blank"
      rel="noopener noreferrer"
    >
      {children}
    </Link>
  );
};

export default ExternalLink;
