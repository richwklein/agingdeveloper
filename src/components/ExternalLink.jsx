import * as React from "react";
import MuiLink from "@mui/material/Link";
import PropTypes from "prop-types";

const ExternalLink = React.forwardRef(function Link(props, ref) {
  return <MuiLink
    href={props.to}
    target="_blank"
    rel="noopener noreferrer"
    ref={ref}
    {...props} />;
});
ExternalLink.propTypes = {
  to: PropTypes.string.isRequired,
};

export default ExternalLink;
