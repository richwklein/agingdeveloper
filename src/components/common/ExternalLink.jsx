import React from "react";
import MuiLink from "@mui/material/Link";
import PropTypes from "prop-types";

// TODO remove spread, fix forwardRef error
const ExternalLink = React.forwardRef((props, ref) => (
  <MuiLink
    data-link="external"
    href={props.to}
    target="_blank"
    rel="noopener noreferrer"
    ref={ref}
    {...props}/>
));

ExternalLink.displayName = "ExternalLink";

ExternalLink.propTypes = {
  to: PropTypes.string.isRequired,
};

export default ExternalLink;
