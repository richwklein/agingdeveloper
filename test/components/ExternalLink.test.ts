import { experimental_AstroContainer as AstroContainer } from 'astro/container';
import { describe, expect, test } from 'vitest';
import ExternalLink from '../../src/components/ExternalLink.astro';

describe("ExternalLink", () => {

  const render = async () => {
    const container = await AstroContainer.create();
    return container.renderToString(ExternalLink, {
      slots: { default: "Test Content" },
      props: { to: "https://example.com" }
    });
  };

  test("that it renders correctly.", async () => {
    const result = await render();
    expect(result).toMatchSnapshot();
  });

  test("that it has a data-link of external.", async () => {
    const result = await render();
    expect(result).toContain("data-link=\"external\"");
  });

  test("that it has a new window target.", async () => {
    const result = await render();
    expect(result).toContain("target=\"_blank\"");
  });

  test("that it has a noopener and noreferrer.", async () => {
    const result = await render();
    expect(result).toContain("rel=\"noopener noreferrer\"");
  });

  test.todo("passing an additional attribute");
});
