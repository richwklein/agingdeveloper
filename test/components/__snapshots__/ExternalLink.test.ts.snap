// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExternalLink > that it renders correctly. 1`] = `"<!DOCTYPE html><!--prettier-ignore --><a data-link="external" target="_blank" rel="noopener noreferrer" href="https://example.com" data-astro-source-file="/Users/richard/Development/Projects/agingdeveloper-astro/src/components/ExternalLink.astro" data-astro-source-loc="20:96">Test Content</a>"`;
